respy/exogenous_processes.py

```diff
@@ -0,0 +1,139 @@
+):
+    """Insert Docstring."""
+    exogenous_processes = optim_paras["exogenous_processes"]
+
+    dense_key_to_exogenous = {
+        key: get_exogenous_from_dense_covariates(
+            dense_key_to_dense_covariates[key], optim_paras
+        )
+        for key in transit_keys
+    }
+
+    # Compute the probabilities for every exogenous process.
+    probabilities = []
+    for exog_proc in exogenous_processes:
+
+        # Create the dot product of covariates and parameters.
+        x_betas = []
+        for params in exogenous_processes[exog_proc].values():
+            x_beta = pandas_dot(states[params.index], params)
+            x_betas.append(x_beta)
+
+        probs = special.softmax(np.column_stack(x_betas), axis=1)
+        probabilities.append(probs)
+
+    # Prepare full Dataframe. If issues arrise we might want to switch typed dicts
+    df = pd.DataFrame(index=states.index)
+
+    for dense in dense_key_to_exogenous:
+        array = functools.reduce(
+            np.multiply,
+            [
+                probabilities[proc][:, val]
+                for proc, val in enumerate(dense_key_to_exogenous[dense])
+            ],
+        )
+        df[str(dense)] = array
+
+    return df
+
+
+@parallelize_across_dense_dimensions
+def weight_continuation_values(
+    complex_, options, continuation_values, transit_key_to_choice_set
+):
+    """Insert Docstring."""
+    transition_df = load_objects("transition", complex_, options)
+
+    choice_set = complex_[1]
+    # Reconsider this setup
+    choice_positons = {
+        key: _get_representation_cols(value, choice_set)
+        for key, value in transit_key_to_choice_set.items()
+    }
+    continuation_values_adjusted = {
+        ftr_key: continuation_values[int(ftr_key)][
+            :, list(choice_positons[int(ftr_key)])
+        ]
+        for ftr_key in transition_df.columns
+    }
+
+    weighted_columns = [
+        transition_df[ftr_key].values.reshape((transition_df.shape[0], 1))
+        * continuation_values_adjusted[ftr_key]
+        for ftr_key in transition_df.columns
+    ]
+
+    continuation_values = functools.reduce(np.add, weighted_columns)
+    return continuation_values
+
+
+def create_transit_choice_set(
+    dense_key_to_transit_representation, dense_key_to_choice_set
+):
+    """Return max representation choice set of each dense choice core."""
+    continuation_choice_sets = {}
+    for dense_key in dense_key_to_transit_representation:
+        continuation_choice_sets[dense_key] = []
+        for transit_representation in dense_key_to_transit_representation:
+            if dense_key in dense_key_to_transit_representation[transit_representation]:
+                continuation_choice_sets[dense_key].append(
+                    dense_key_to_choice_set[transit_representation]
+                )
+    out = {}
+    for dense_key in continuation_choice_sets:
+        out[dense_key] = _get_maximal_choice_set(continuation_choice_sets[dense_key])
```

Comment:
I think this is true but does not necessarily contribute to a more readable code in this situation since the dict comp will be quite complicated. What is your opinion?

Reply:
It's fine for me the way it is! I agree that the loop is probably more readable. It just seemed to me like we are going for comprehensions whenever possible (I thought they are sometimes a bit faster but I might be misinformed there) so just wanted to point it out.